### PR TITLE
Add pip install with editable mode

### PR DESCRIPTION
**Pull request by adam2392**
Closes: #138

### CODE REVIEW

**Comment by adam2392** [comment]
Just want to clarify that this is moving in the right direction?

File: spin/__main__.py [hunk 1]
```diff
@@ -87,6 +87,7 @@ def group(ctx):
         "spin.ipython": _cmds.meson.ipython,
         "spin.python": _cmds.meson.python,
         "spin.shell": _cmds.meson.shell,
+        "pip.isntall": _cmds.meson.install,
```
Comment (stefanv) [inline spin/__main__.py:90]:
```suggestion
```

File: spin/cmds/meson.py [hunk 2]
```diff
@@ -601,3 +601,29 @@ def docs(ctx, sphinx_target, clean, first_build, jobs, sphinx_gallery_plot):
         f"$ export PYTHONPATH={os.environ['PYTHONPATH']}", bold=True, fg="bright_blue"
     )
     _run(["make", "-C", "doc", sphinx_target], replace=True)
+
+
+@click.command()
```
Comment (stefanv) [inline spin/cmds/meson.py:606]: Place this command in a new file, called `pip.py`

File: spin/cmds/meson.py [hunk 3]
```diff
@@ -601,3 +601,29 @@ def docs(ctx, sphinx_target, clean, first_build, jobs, sphinx_gallery_plot):
...
+@click.option("-j", "--jobs", help="Number of parallel tasks to launch", type=int)
+@click.option("--clean", is_flag=True, help="Clean build directory before build")
```
Comment (stefanv) [inline spin/cmds/meson.py:608]:
```suggestion
```

File: spin/cmds/meson.py [hunk 4]
```diff
@@ -601,3 +601,29 @@ def docs(ctx, sphinx_target, clean, first_build, jobs, sphinx_gallery_plot):
...
+@click.option(
+    "-v", "--verbose", is_flag=True, help="Print detailed build and installation output"
+)
+@click.argument("-e", "--editable", is_flag=True, help="Install in editable mode")
+@click.argument("meson_args", nargs=-1)
```
Comment (stefanv) [inline spin/cmds/meson.py:613]:
```suggestion
@click.argument("pip_args", nargs=-1)
```

File: spin/cmds/meson.py [hunk 5]
```diff
@@ -601,3 +601,29 @@ def docs(ctx, sphinx_target, clean, first_build, jobs, sphinx_gallery_plot):
...
+@click.pass_context
```
Comment (stefanv) [inline spin/cmds/meson.py:614]:
```suggestion
```

File: spin/cmds/meson.py [hunk 6]
```diff
@@ -601,3 +601,29 @@ def docs(ctx, sphinx_target, clean, first_build, jobs, sphinx_gallery_plot):
...
+    """💽 Install package and also build.
+
+    MESON_ARGS are passed through in the same manner as the build function e.g.:
+
+      spin install -- -Dpkg_config_path=/lib64/pkgconfig
+
+    The package can also be installed in editable mode.
+    """
+    ctx.invoke(build, meson_args=meson_args, jobs=jobs, clean=clean, verbose=verbose, quiet=quiet)
+
```
Comment (stefanv) [inline spin/cmds/meson.py:625]:
```suggestion
```

File: spin/cmds/meson.py [hunk 7]
```diff
@@ -601,3 +601,29 @@ def docs(ctx, sphinx_target, clean, first_build, jobs, sphinx_gallery_plot):
...
+    if editable:
+        _run(["pip", "install", "--no-build-isolation", "--editable", "."], replace=True)
```
Comment (stefanv) [inline spin/cmds/meson.py:627]: add pip args

**Review by stefanv — COMMENTED** [review]
I've left some comments, but this is pretty much the gist of it, yes!

File: spin/cmds/meson.py (hunk 2 above)
Comment (adam2392) [reply spin/cmds/meson.py:606]: Done

File: spin/cmds/pip.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+from .util import run as _run
+
+
+@click.command()
+@click.option(
+    "-v", "--verbose", is_flag=True, help="Print detailed build and installation output"
+)
+@click.argument("-e", "--editable", is_flag=True, help="Install in editable mode")
+@click.argument("pip_args", nargs=-1)
+def install(pip_args, verbose=False, editable=True, quiet=False):
```
Comment (stefanv) [inline spin/cmds/pip.py:12]:
```suggestion
def install(pip_args, verbose=False, editable=True):
```

File: spin/cmds/pip.py [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    """💽 Install package and also build.
+
+    pip arguments are passed through in the same manner as you would pip e.g.:
+
+      spin install -- -r requirements.txt
```
Comment (stefanv) [inline spin/cmds/pip.py:17]: Would this example make sense? I'd read it as:

```
pip install . --editable --no-build-isolation -r requirements.txt
```

which doesn't make so much sense.

File: spin/cmds/pip.py [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+    if editable:
+        setup_cmd = [
+            "pip",
+            "install",
+            ".",
+            "--no-build-isolation",
+            "--editable",
+        ] + list(pip_args)
+    else:
+        setup_cmd = ["pip", "install", "."] + list(pip_args)
```
Comment (stefanv) [inline spin/cmds/pip.py:31]:
```suggestion
    pip_args = list(pip_args)
    pip_cmd = ["pip", "install", "."]
    if editable:
        pip_args += ["--no-build-isolation", "--editable"]
```

File: spin/cmds/pip.py [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+
+    verbose_flag = ["-v"] if verbose else []
+    setup_cmd += verbose_flag
```
Comment (stefanv) [inline spin/cmds/pip.py:34]:
```suggestion
    pip_cmd += ["-v"] if verbose else []
```

File: spin/cmds/pip.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+            "pip",
+            "install",
+            ".",
+            "--no-build-isolation",
+            "--editable",
+        ] + list(pip_args)
+    else:
+        setup_cmd = ["pip", "install", "."] + list(pip_args)
+
+    verbose_flag = ["-v"] if verbose else []
```
Comment (stefanv) [inline spin/cmds/pip.py:33]:
```suggestion
```

File: spin/cmds/pip.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    setup_cmd += verbose_flag
+
+    p = _run(setup_cmd, sys_exit=False, output=not quiet)
```
Comment (stefanv) [inline spin/cmds/pip.py:36]:
```suggestion
    p = _run(pip_cmd, sys_exit=False, replace=True)
```

File: spin/cmds/pip.py [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+    if p.returncode != 0:
+        raise RuntimeError("Pip installation failed; please try `spin install` again.")
```
Comment (stefanv) [inline spin/cmds/pip.py:38]:
```suggestion
```

I think pip will give helpful enough feedback here.

File: spin/cmds/pip.py (hunk 9 above)
Comment (adam2392) [reply spin/cmds/pip.py]: Ah okay, I clarified and edited slightly. With the new logic of the `install` function, hopefully this is better?

File: spin/cmds/pip.py [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    "-v", "--verbose", is_flag=True, help="Print detailed build and installation output"
+)
+@click.argument("-e", "--editable", is_flag=True, help="Install in editable mode")
+@click.argument("pip_args", nargs=-1)
+def install(pip_args, verbose=False, editable=True):
+    """💽 Install package and also build.
+
+    pip arguments are passed through in the same manner as you would pip e.g.:
+
+        spin install -- -r requirements.txt --editable false
```
Comment (stefanv) [inline spin/cmds/pip.py:17]: My point was that you probably don't want to do this. You want something like: `spin install -- --no-clean` or another option that makes sense

File: spin/cmds/pip.py [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+        pip install -r requirements.txt
+
+    The package is by default installed in editable mode, but you can disable this
+    by passing the `--editable` flag.
+    """
+    pip_args = list(pip_args)
+    pip_cmd = ["pip", "install", "."]
+    if editable:
+        pip_args += ["--no-build-isolation", "--editable"]
```
Comment (stefanv) [inline spin/cmds/pip.py:29]: Maybe `--no-build-isolation` should be used in both cases? I.e.,`pip_cmd = ["pip", "install", ".", "--no-build-isolation"]`

**Review by stefanv — COMMENTED** [review]
Looking good; you should also add this to the CI test.sh to ensure that it does what you expect it to.

**Comment by stefanv** [comment]
And to `pyproject.toml` of the example project.

File: spin/cmds/pip.py (hunk 16 above)
Comment (adam2392) [reply spin/cmds/pip.py:29]: Sure fair.

File: spin/cmds/pip.py (hunk 15 above)
Comment (adam2392) [reply spin/cmds/pip.py:17]: Ah I see. sure that makes sense. I edited the example

**Comment by adam2392** [comment]
Few questions:

1. Should I add a changelog
2. Is there a documented way to test this locally?

**Comment by stefanv** [comment]
Thanks @adam2392!

**Pull request merged**